### PR TITLE
Jetpack: update react-router-dom from v5 to v6

### PR DESCRIPTION
**Pull request by tbradsha**
This is PR bumps Jetpack-the-plugin to use `react-router-dom` to v6 (and ultimately to v7).

## Proposed changes:
Essentially I followed this guide:
https://reactrouter.com/6.28.0/upgrading/v5

Of note:
* `<Prompt>` is no longer supported in v6, so I've commented the one instance out.
* Several `withRouter` instances were updated. These really should have been upgraded prior to moving to v5.1 (e.g. in #15514 or #15821), but better late than never.
* One instance of the now-obsolete `history.listen` was updated (h/t @anomiex for a solution here); see inline comment.

That last item could probably use the most scrutiny in a code review.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* Verify a clickthrough works as expected.
* Verify CI is happy.

### CODE REVIEW

File: projects/plugins/jetpack/_inc/client/components/navigation-settings/index.jsx [hunk 1]
```diff
@@ -37,9 +37,14 @@ export class NavigationSettings extends React.Component {
...
-		// We need to handle the search term not only on route update but also on page load in case of some external redirects
 		this.onRouteChange( this.props.location );
-		this.props.history.listen( this.onRouteChange );
+	}
+
+	componentDidUpdate( oldprops ) {
+		// We need to handle the search term not only on route update but also on page load in case of some external redirects
+		if ( oldprops.location !== this.props.location ) {
+			this.onRouteChange( this.props.location );
+		}
```
Comment (tbradsha) [inline projects/plugins/jetpack/_inc/client/components/navigation-settings/index.jsx]: There's no direct upgrade path for this, but hopefully this roughly the same thing as before.

File: projects/plugins/jetpack/_inc/client/components/jetpack-notices/plan-conflict-warning.jsx [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 import SimpleNotice from 'components/notice';
 import { getActiveSitePurchases } from 'state/site';
 
 /**
  * PlanConflictWarning component
  *
+ * @param {object} root0                     - props
+ * @param {Array}  root0.activeSitePurchases - active site purchases
+ * @param {object} root0.location            - location object
+ * @param {string} root0.location.pathname   - location pathname
```
Comment (anomiex) [inline projects/plugins/jetpack/_inc/client/components/jetpack-notices/plan-conflict-warning.jsx]: Instead of bc034371f943b0cc763b6254ccd7d887b9ee88ea, it looks like a more "proper" way to fix the tests may be to put them inside a `<MemoryRouter>` that is initialized with the correct location paths, something like this
```jsx
<MemoryRouter initialEntries={ [ '/whatever' ] }>
    <PlanConflictWarning />
</MemoryRouter>
```
Although to get it to actually work I wound up with something more like this:
```jsx
/**
 * Wrap the component in a `<MemoryRouter>` for testing.
 *
 * @param {PlanConflictWarning} component The `<PlanConflictWarning>` to wrap.
 * @param {string[]} entries Initial path entries.
 * @returns {MemoryRouter} Router to render..
 */
function wrapWithMemoryRouter( component, entries = [ '/plans' ] ) {
	return (
		<MemoryRouter
			initialEntries={ entries }
			future={
				// Enable some forward-compat things so it doesn't console.warn about enabling them. Sigh.
				{ v7_startTransition: true, v7_relativeSplatPath: true }
			}
		>
			{ component }
		</MemoryRouter>
	);
}
```

File: projects/plugins/jetpack/_inc/client/components/navigation-settings/index.jsx [hunk 3]
```diff
@@ -37,9 +37,14 @@ export class NavigationSettings extends React.Component {
 	};
 
 	UNSAFE_componentWillMount() {
-		// We need to handle the search term not only on route update but also on page load in case of some external redirects
 		this.onRouteChange( this.props.location );
-		this.props.history.listen( this.onRouteChange );
+	}
+
+	componentDidUpdate( oldprops ) {
+		// We need to handle the search term not only on route update but also on page load in case of some external redirects
```
Comment (anomiex) [inline projects/plugins/jetpack/_inc/client/components/navigation-settings/index.jsx]: The bits here on line 45–47 are the "on route update", while the `UNSAFE_componentWillMount()` is the "also on page load". I think this comment may be in the wrong place?

File: projects/plugins/jetpack/_inc/client/main.jsx [hunk 4]
```diff
@@ -854,10 +857,12 @@ class Main extends React.Component {
 					<AdminNotices />
 					<JetpackNotices />
 					{ this.shouldConnectUser() && this.connectUser() }
+					{ /* This is no longer supported as of react-router-dom v6: https://github.com/remix-run/react-router/issues/8139
```
Comment (anomiex) [inline projects/plugins/jetpack/_inc/client/main.jsx]: If we really want it, I think we could use [`unstable_usePrompt()`](https://api.reactrouter.com/v7/functions/react_router.unstable_usePrompt.html). It claims it can be weird in different browsers in some cases though.

For use inside a class component, like here, that might mean creating our own tiny `<Prompt>` component something like this:
```jsx
function Prompt( { when, message } ) {
	unstable_usePrompt( { when, message } );
	return null;
}
```

**Comment by anomiex** [comment]
I think I see why the E2Es are failing.

Formerly, due to lack of `exact`, https://github.com/Automattic/jetpack/blob/f66d30338b34193ded5e1e42fd0e04023ceb5034/projects/plugins/jetpack/_inc/client/admin.js#L56-L58 would match anything beginning with `/recommendations`. But that was ok, since https://github.com/Automattic/jetpack/blob/f66d30338b34193ded5e1e42fd0e04023ceb5034/projects/plugins/jetpack/_inc/client/recommendations/index.jsx#L214-L334 uses absolute paths.

Now, though, we get the `/recommandations` route failing but the rest like `/recommandations/site-type` succeeding due to two things:

1. Since it doesn't end with `/*`, the route at https://github.com/Automattic/jetpack/blob/bc034371f943b0cc763b6254ccd7d887b9ee88ea/projects/plugins/jetpack/_inc/client/admin.js#L51-L54 only matches `/recommandations`, not anything like `/recommandations/site-type`. The route at https://github.com/Automattic/jetpack/blob/bc034371f943b0cc763b6254ccd7d887b9ee88ea/projects/plugins/jetpack/_inc/client/recommendations/index.jsx#L216-L219 fails to match it, because since paths are relative now it's trying to match for `/recommandations/recommendations` instead.
2. `/recommandations/site-type` and the like continue to work because they go via https://github.com/Automattic/jetpack/blob/bc034371f943b0cc763b6254ccd7d887b9ee88ea/projects/plugins/jetpack/_inc/client/admin.js#L90 instead.

AFAICT, to fix all this we need to make that be `path="/recommendations/*"` in [the first bit](https://github.com/Automattic/jetpack/blob/bc034371f943b0cc763b6254ccd7d887b9ee88ea/projects/plugins/jetpack/_inc/client/admin.js#L51-L54) and remove the "/recommendations" prefix from everything in [the second bit](https://github.com/Automattic/jetpack/blob/bc034371f943b0cc763b6254ccd7d887b9ee88ea/projects/plugins/jetpack/_inc/client/recommendations/index.jsx#L216-L360) (which means that first entry will become `path=""`, and the rest can apparently be either like `path="site-type"` or `path="/site-type"`).

**Comment by anomiex** [comment]
P.S. Something that's really confusing is that v6 allows absolute paths in nested routes for something like
```jsx
<Routes>
  <Route path="/foo/*" element={<Foo />}>
    <Route path="/foo/bar" element={<FooBar />} />
  </Route>
</Routes>
```
but doesn't for a case like ours which looks more like
```jsx
<Routes>
  <Route path="/foo/*" element={<Foo />} />
</Routes>
```
with `Foo` containing
```jsx
<Routes>
  <Route path="/foo/bar" element={<FooBar />} />
</Routes>
```
So until you come across https://github.com/remix-run/react-router/issues/10321 / https://github.com/remix-run/react-router/discussions/9841 that explicitly point this out, you're likely to find confusing results talking about it working because they only consider the first case.

File: projects/plugins/jetpack/_inc/client/components/jetpack-notices/plan-conflict-warning.jsx (hunk 2 above)
Comment (tbradsha) [reply projects/plugins/jetpack/_inc/client/components/jetpack-notices/plan-conflict-warning.jsx]: Stole your function, comments and all: 06f4482fc3

File: projects/plugins/jetpack/_inc/client/components/navigation-settings/index.jsx (hunk 3 above)
Comment (tbradsha) [reply projects/plugins/jetpack/_inc/client/components/navigation-settings/index.jsx]: Oops: 785a6433e2

File: projects/plugins/jetpack/_inc/client/main.jsx (hunk 4 above)
Comment (tbradsha) [reply projects/plugins/jetpack/_inc/client/main.jsx]: Good call; looks like it still works on some cases (see #15514). I'll reimplement with the above.

File: projects/plugins/jetpack/_inc/client/components/jetpack-notices/plan-conflict-warning.jsx (hunk 2 above)
Comment (anomiex) [reply projects/plugins/jetpack/_inc/client/components/jetpack-notices/plan-conflict-warning.jsx]: That's why I included it, in case you wanted to steal it. 🙂

**Comment by anomiex** [comment]
In manual testing, all the routes other than the recommendations stuff seem to work. 👍 Waiting on the build to test the new `<Prompt>` thing.

File: projects/plugins/jetpack/_inc/client/main.jsx (hunk 4 above)
Comment (tbradsha) [reply projects/plugins/jetpack/_inc/client/main.jsx]: 86ffa45

**Comment by tbradsha** [comment]
> Waiting on the build to test the new `<Prompt>` thing.

I had a hard time figuring out where to test this, but in the end went to Settings → Discussion, toggled on "Comments", and modified the "Comment form introduction" text field, then tried to navigate away:

`/wp-admin/admin.php?page=jetpack#discussion`

It worked in `trunk`, but not in this branch...

File: projects/plugins/jetpack/_inc/client/main.jsx (hunk 4 above)
Comment (tbradsha) [reply projects/plugins/jetpack/_inc/client/main.jsx]: But it seems to not work still. 🤔

**Comment by anomiex** [comment]
It gives me a blank screen in JN. Compiling a dev build locally, I get an error about "useBlocker must be used within a data router". 

Once I jump through hoops to fix that (apparently instead of `<HashRouter>` it needs to use `createHashRouter()` and `<RouterProvider>`, see [docs](https://reactrouter.com/6.28.1/routers/picking-a-router) and [SO](https://stackoverflow.com/questions/75135147/react-router-dom-v6-useblocker)), it still doesn't prompt, and I get console messages like
> You are trying to use a blocker on a POP navigation to a location that was not created by @remix-run/router. This will fail silently in production. This can happen if you are navigating outside the router via `window.history.pushState`/`window.location.hash` instead of using router navigation APIs.  This can also happen if you are using createHashRouter and the user manually changes the URL.

In other words, it seems borked if you use actual links for navigation. 🙄 At this point I'm inclined to punt, revert https://github.com/Automattic/jetpack/pull/40773/commits/86ffa45c8c7dea1470b8e21a4b8371a9d96bf4fb and just update the comment saying we might use `unstable_usePrompt` if we want to bring that back but it has warnings about odd behavior and also issues with HashRouter that would need resolving.

File: projects/plugins/jetpack/_inc/client/main.jsx [hunk 5]
```diff
@@ -854,10 +857,15 @@ class Main extends React.Component {
 					<AdminNotices />
 					<JetpackNotices />
 					{ this.shouldConnectUser() && this.connectUser() }
+					{ /*
+					This component was removed as of react-router-dom v6: https://github.com/remix-run/react-router/issues/8139
+					It could probably be brought back with `unstable_usePrompt`, but that would require jumping through hoops with
+					createHashRouter(), etc., and is already not reliable cross-browser anyway.
```
Comment (anomiex) [inline projects/plugins/jetpack/_inc/client/main.jsx]: More important than the need to switch from `<HashRouter>` to `createHashRouter()` is that it doesn't even work then anyway.
```suggestion
					It could probably be brought back with `unstable_usePrompt`, but that is broken with the hash router and normal links,
					and is already not reliable cross-browser anyway.
```

**Review by anomiex — APPROVED** [review]
E2Es are happy, recommendations seem happy, let's do it.
